### PR TITLE
Updates HasOptions to account for initialOptions

### PR DESCRIPTION
**Pull request by jeremeamia**
Updates `HasOptions` to account for `initialOption(s)`, including providing validation and array-ifying helpers. Introduces `OptionsConfig` as a way to parameterize `HasOptions` behaviors. Also, updates all `HasOptions`/`HasOptionGroups`-using elements. Breaks `isInitial` back out from the `Option` as a separately-managed thing.

In addition, adds `setExtra` to base `Element` object to allow for patching on new Slack features in a non-conflicting way prior to the library being updated. Also, updates some other menus with new/missing parameters.

### CODE REVIEW

**Comment by jeremeamia** [comment]
@Andi1986 I would like to make these changes prior to releasing. Mind giving these a look?

File: src/Partials/Option.php [hunk 1]
```diff
@@ -14,17 +17,20 @@ class Option extends Element
...
-        $this->value($value);
-        $this->isInitial = $isInitial;
+    /**
+     * @param string $text
+     * @param string $value
+     * @return Option
+     */
+    public static function create(string $text, string $value): self
+    {
+        return Option::new()->text($text)->value($value);
```
Comment (Andi1986) [inline src/Partials/Option.php]: just out of curiousity, whats the benefit compared to simple "new" here ? text and value could be set in constructor ?

File: src/Inputs/OverflowMenu.php [hunk 2]
```diff
@@ -4,42 +4,48 @@
...
+    }
 
-    public function validate(): void
+    /**
+     * @param string $text
+     * @param string $value
+     * @param string $url
+     * @return static
+     */
+    public function urlOption(string $text, string $value, string $url): self
```
Comment (Andi1986) [inline src/Inputs/OverflowMenu.php:28]: interesting, how did u find this ? it was not documented https://api.slack.com/reference/block-kit/block-elements#overflow

File: src/Element.php [hunk 3]
```diff
@@ -59,7 +82,15 @@ public function toArray(): array
         $this->validate();
         $type = $this->getType();
 
-        return !in_array($type, Type::HIDDEN_TYPES, true) ? compact('type') : [];
+        $data = !in_array($type, Type::HIDDEN_TYPES, true) ? compact('type') : [];
+
+        if (!empty($this->extra)) {
+            foreach ($this->extra as $key => $value) {
```
Comment (Andi1986) [inline src/Element.php]: could remove the if with:
foreach ($this->extra ?? [] as $key => $value) {

File: src/Partials/HasOptions.php [hunk 4]
```diff
@@ -56,4 +169,23 @@ protected function getOptionsAsArray(): array
...
+        $data = [];
+
+        if (empty($this->initialOptions)) {
+            return $data;
+        }
+
+        $maxInitialOptions = (int) $this->config()->getMaxInitialOptions();
+
+        if ($maxInitialOptions === 1) {
+            return $data + ['initial_option' => $this->initialOptions[0]->toArray()];
```
Comment (Andi1986) [inline src/Partials/HasOptions.php]: the $data value is always empty. should be removed i guess

- [ ] $data removed

**Review by Andi1986 — APPROVED** [review]
nice refactoring of options! makes sense as they appear in alot of elements

File: src/Inputs/OverflowMenu.php (hunk 2 above)
Comment (jeremeamia) [reply src/Inputs/OverflowMenu.php:28]: Found it in the Option docs: https://api.slack.com/reference/block-kit/composition-objects#option

File: src/Partials/Option.php (hunk 1 above)
Comment (jeremeamia) [reply src/Partials/Option.php]: I had a reason while I was in the middle of doing it, but apparently that reason wasn't very necessary by the end. Thanks for pointing this out so I can go back to `new`.

**Pull request merged**